Add use case for getting dataset versions

vera opened the pull request:
## What this PR does / why we need it:

This PR adds code for using the ["List Versions of a Dataset" API](https://guides.dataverse.org/en/latest/api/native-api.html#list-versions-of-a-dataset).

## Which issue(s) this PR closes:

I'm not aware of an open issue.

## Related Dataverse PRs:

- ~Depends on https://github.com/IQSS/dataverse/pull/12008~

Since during review of https://github.com/IQSS/dataverse/pull/12008, it was decided not to add the `isInReviewState` flag to dataset versions API responses, I've removed them from the code here as well. This PR no longer depends on/is related to any dataverse PRs.

## Special notes for your reviewer:

/

## Suggestions on how to test this:

I've added tests and docs on how to use.

## Is there a release notes or changelog update needed for this change?:

Yes

## Additional documentation:

/

vera committed:
feat: add use case for getting dataset versions

vera committed:
docs: add use case for getting dataset versions

vera committed:
test: add use case for getting dataset versions

ekraffmiller commented on docs/useCases.md:
```diff
@@ -871,6 +872,36 @@ Note that `collectionId` is an optional parameter to filter datasets by collecti
+- **limit**: (number) Limit for pagination.
+- **offset**: (number) Offset for pagination.
+- **excludeMetadataBlocks**: (boolean) Exclude metadata blocks (default: false).
+- **excludesFiles**: (boolean) Exclude files (default: true).
```
I think returning the complete list of files in the response could be a problem - some of our datasets have several thousand files.  We have another use case, GetDatasetFiles, which returns the files with pagination.  My suggestion is to remove files from the response, and use the other use case to get the files with pagination.

vera replied:
@ekraffmiller thanks for your review and sorry for the delay in getting back to this.

I have removed the `excludesFiles` param from this client implementation, so it's no longer possible to request the list of files without pagination.

I'm just wondering:

* Should I also add a note in the documentation of the use case that we are differing from the Dataverse API specification here, and why?
* Does this mean that the param should really be removed from the Dataverse API entirely? Or are there other use cases, outside of dataverse-client-javascript, which won't have this problem?

ekraffmiller replied:
I think it's ok to differ from the Dataverse API, because it is similar to what we are doing in the GetDataset use case.  We have a separate use case, getDatasetFiles, that gets the files for a dataset version, with pagination.  It could be helpful, though to put a comment in the docs about that, so users know to use getDatasetFiles.  
We might want to remove this from the Dataverse API, but it will be a breaking change, so will have to think about that more.

vera replied:
OK, I've added a comment to the docs and linked the getDatasetFiles use case.

ekraffmiller reviewed (changes requested):
Hi @vera, thanks for your PR. We don't currently need this use case in our SPA, because we use GetDatasetVersionsSummaries, and GetDatasetVersionDiff to display a list of Dataset versions.  But since it is available on the Dataverse API I think it's ok to add.  The only caveat I have is to not return the list of files, for performance reasons.

vera committed:
fix: remove "isInReviewState" flag from dataset versions

vera committed:
fix: remove "excludeFiles" flag from "Get Dataset Versions" use case

vera committed:
fix: update GetDatasetVersions.test.ts

ekraffmiller commented on src/datasets/domain/models/DatasetVersion.ts:
```diff
@@ -0,0 +1,30 @@
+  citationDate: string
+  license: DatasetLicense
+  fileAccessRequest: boolean
+  files?: Array<FilePayload>
```
```suggestion
```
Since we are not returning files, this field can be removed from the model.

ekraffmiller commented on test/integration/datasets/DatasetsRepository.test.ts:
```diff
@@ -1555,6 +1555,200 @@ describe('DatasetsRepository', () => {
+      expect(allVersions.totalCount).toBe(22)
+
+      await deletePublishedDatasetViaApi(testDatasetIds.persistentId)
+    }, 180000)
```
```suggestion
      await deletePublishedDatasetViaApi(testDatasetIds.persistentId)
    }, 5000)
```
Can we shorten the time for this test? Fewer pages or fewer items per page?

vera replied:
Sure. I had copied the number of versions to create and the timeout from another test ('should return dataset versions summaries with pagination'). I'm not sure if the 5000 timeout is sufficient for the test now, should I decrease any other numbers in the test?

I tried to check by running `npm run test:integration -- -t "getDatasetVersions"`, but it seems I'm unable to run the integration tests locally, because I get some unrelated error about a missing storage driver API:

```
            message: 'API endpoint does not exist on this server. Please check your code for typos, or consult our API guide at http://guides.dataverse.org.',
            requestUrl: 'http://localhost:8080/api/v1/admin/dataverse/testDatasetVersionsCollection/storageDriver',
            requestMethod: 'PUT'
          }
        },
        status: 404
      }

      at test/testHelpers/collections/collectionHelper.ts:144:13
          at Generator.throw (<anonymous>)

  ● DatasetsRepository › getDatasetVersions › should return dataset versions when dataset exists

    Error while setting storage driver for collection testDatasetVersionsCollection

      143 |   } catch (error) {
      144 |     console.log(error)
    > 145 |     throw new Error(`Error while setting storage driver for collection ${collectionAlias}`)
          |           ^
      146 |   }
      147 | }
      148 |

      at test/testHelpers/collections/collectionHelper.ts:145:11
          at Generator.throw (<anonymous>)
      at rejected (test/testHelpers/collections/collectionHelper.ts:6:65)
```

vera committed:
docs: add note about missing excludeFiles parameter in "Get Dataset Versions" use case

vera committed:
fix: apply suggestions from code review

Co-authored-by: Ellen Kraffmiller <ekraffmiller@g.harvard.edu>

vera committed:
Merge remote-tracking branch 'refs/remotes/vera/feat/get-dataset-versions' into feat/get-dataset-versions

ekraffmiller reviewed (approved):
looks good, thanks for the updates

vera committed:
Merge remote-tracking branch 'origin/develop' into feat/get-dataset-versions

vera committed:
docs: update CHANGELOG.md to include `getDatasetVersions` use case

vera committed:
fix: remove `totalCount` from `getDatasetVersions` and related tests

vera commented:
> Hi @vera , can you update this PR with the latest from develop? I think that should fix the integration test.

@ekraffmiller I've updated the PR with develop, and also fixed some other mistakes which were causing the integration tests to fail (1. removed the `totalCount`, which I wrongly assumed the Dataverse API returns, but it doesn't actually do that, and 2. increased the timeout for the pagination test with 21 versions, because it had previously failed due to a timeout: https://github.com/IQSS/dataverse-client-javascript/actions/runs/32116134773/job/95645950539#step:5:456)

The tests are succeeding now.

I've also added a line to the changelog.

ChengShi-1 commented:
Hi Vera, since currently the `GetVersion` api has a bug flagged in https://github.com/IQSS/dataverse.harvard.edu/issues/397 and we are not using it now for Modern UI, I'll put this On Hold untill we solve the bug
